### PR TITLE
[BE] - Following api

### PR DESCRIPTION
**Pull request by sonrhey27**
### Asana: 
[link](https://app.asana.com/0/1202070327408572/board)

### Definition of Done:  
[BE] - Following API

### Notes:
Postman [link](https://go.postman.co/workspace/My-APIs~3f417729-0358-40b4-8ee5-0e22089685de/collection/20198092-665b0396-e50d-40f7-a5ee-12b67f10cd81?action=share&creator=20198092) for reference

### Screenshots:

image

### CODE REVIEW

File: backend/app/Http/Controllers/FollowingController.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            $followed_back = UserRelationshipModel::find($usra->user_relationship_id);
+            $followed_back->followed_id = $my_id;
+            $followed_back->save();
+
+            $following = new UserRelationshipModel();
+            $following->following_id = $request->follow_back_id;
+            $following->followed_id = $request->follow_back_id;
+            $following->save();
+
+            $user_activities = new UserActivitiesModel();
```
Comment (inahsanjuan) [inline backend/app/Http/Controllers/FollowingController.php]: This is not included in our scope.
So please remove code related to UserActivity

File: backend/app/Http/Controllers/FollowingController.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            DB::rollback();
+            $this->response->status_code = 0;
+            $this->response->message = "error";
+            $this->response->data = "Data Error";
+
+            return response()->json($this->response);
+        }
+    }
+
+    public function follow_back(Request $request){
```
Comment (inahsanjuan) [inline backend/app/Http/Controllers/FollowingController.php]:
```suggestion
    public function follow_back(Request $request) {
```

File: backend/app/Http/Controllers/FollowingController.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    public function get_user_relationship(){
+        $usr = UserRelationshipModel::all();
+        $this->response->status_code = 1;
+        $this->response->message = "success";
+        $this->response->data = $usr;
+
+        return response()->json($this->response);
+    }
+
+    public function follow_user(Request $request){
```
Comment (inahsanjuan) [inline backend/app/Http/Controllers/FollowingController.php]:
```suggestion
    public function follow_user(Request $request) {
```

File: backend/app/Http/Controllers/FollowingController.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+use App\Models\User;
+
+class FollowingController extends Controller
+{
+    private $response;
+    public function __construct() {
+        $this->response = new Response();
+    }
+
+    public function get_user_activities(){
```
Comment (inahsanjuan) [inline backend/app/Http/Controllers/FollowingController.php]:
```suggestion
    public function get_user_activities() {
```

File: backend/app/Http/Controllers/FollowingController.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        $usra = UserActivitiesModel::with('user_relationship.following','user_relationship.followed_back', 'user')->get();
+        $this->response->status_code = 1;
+        $this->response->message = "success";
+        $this->response->data = $usra;
+
+        return response()->json($this->response);
+    }
+
+    public function get_user_relationship(){
```
Comment (inahsanjuan) [inline backend/app/Http/Controllers/FollowingController.php]:
```suggestion
    public function get_user_relationship() {
```

File: backend/app/Http/Controllers/FollowingController.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            DB::rollback();
+            $this->response->status_code = 0;
+            $this->response->message = "error";
+            $this->response->data = "Data Error";
+
+            return response()->json($this->response);
+        }
+    }
+
+    public function user_list(){
```
Comment (inahsanjuan) [inline backend/app/Http/Controllers/FollowingController.php]:
```suggestion
    public function user_list() {
```

File: frontend/src/components/ProfileComponent.vue [hunk 7]
```diff
@@ -10,43 +10,142 @@
...
+        })
+        const response = await submitRequest.data
+        if (response.status_code === 1) {
+            csvc.updateUser(response.data)
+            location.reload()
+            return
+        }
+        errorMessage.value = response.data
+      } catch(e) {
+        console.log(e.response)
```
Comment (inahsanjuan) [inline frontend/src/components/ProfileComponent.vue]:
```suggestion
        console.error(e.response)
```

File: frontend/src/components/ProfileComponent.vue [hunk 8]
```diff
@@ -10,43 +10,142 @@
...
+        'image' : null
+      }
+
+      const data = editedField(inputStatus, inputValues)
+      const formData = new FormData()
+      console.log(data);
+      formData.append('fields', JSON.stringify(data))
+      formData.append('image', avatar.value)
+
+      try{
```
Comment (inahsanjuan) [inline frontend/src/components/ProfileComponent.vue]:
```suggestion
      try {
```

File: frontend/src/components/ProfileComponent.vue [hunk 9]
```diff
@@ -10,43 +10,142 @@
...
+        'full_name' : full_name.value.value,
+        'email_address' : email_address.value.value,
+        'password' : password.value.value,
+        'new_password' : new_password.value,
+        'image' : null
+      }
+
+      const data = editedField(inputStatus, inputValues)
+      const formData = new FormData()
+      console.log(data);
```
Comment (inahsanjuan) [inline frontend/src/components/ProfileComponent.vue]: Remove unnecessary console.log

**Review by inahsanjuan — CHANGES_REQUESTED** [review]

File: backend/app/Http/Controllers/FollowingController.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        $usr = UserRelationshipModel::all();
+        $this->response->status_code = 1;
+        $this->response->message = "success";
+        $this->response->data = $usr;
+
+        return response()->json($this->response);
+    }
+
+    public function follow_user(Request $request) {
+        try{
```
Comment (inahsanjuan) [inline backend/app/Http/Controllers/FollowingController.php]:
```suggestion
        try {
```

File: backend/app/Models/UserActivitiesModel.php [hunk 11]
```diff
@@ -16,4 +16,12 @@ class UserActivitiesModel extends Model
...
         'user_id'
     ];
+
+    public function user(){
+        return $this->belongsTo(User::class, 'user_id', 'user_id');
+    }
+
+    public function user_relationship(){
+        return $this->hasMany(UserRelationshipModel::class, 'user_relationship_id', 'user_relationship_id');
+    }
```
Comment (inahsanjuan) [inline backend/app/Models/UserActivitiesModel.php]:
```suggestion
    public function user() {
        return $this->belongsTo(User::class, 'user_id', 'user_id');
    }

    public function user_relationship() {
        return $this->hasMany(UserRelationshipModel::class, 'user_relationship_id', 'user_relationship_id');
    }
```

File: backend/database/migrations/2022_04_06_025951_create_user_relationship_models_table.php [hunk 12]
```diff
@@ -15,8 +15,8 @@ public function up()
     {
         Schema::create('user_relationships', function (Blueprint $table) {
             $table->bigIncrements('user_relationship_id');
-            $table->integer('following_id');
-            $table->integer('followed_id');
+            $table->integer('following_id')->nullable();
+            $table->integer('followed_id')->nullable();
```
Comment (inahsanjuan) [inline backend/database/migrations/2022_04_06_025951_create_user_relationship_models_table.php]: I think its better if this is not nullable?
In which scenario will this be applicable if ever

**Review by inahsanjuan — CHANGES_REQUESTED** [review]

File: backend/database/migrations/2022_04_06_025951_create_user_relationship_models_table.php (hunk 12 above)
Comment (sonrhey27) [reply backend/database/migrations/2022_04_06_025951_create_user_relationship_models_table.php]: I will edit this one, I was just overthinking a scenario and looks like I don't need it to be nullable

**Pull request merged**